x-pack/platform/plugins/shared/alerting/server/alerts_client/alerts_client.ts

```diff
@@ -158,43 +168,24 @@ export class AlertsClient<
     }
     await this.legacyAlertsClient.initializeExecution(opts);
 
-    if (!this.ruleType.alerts?.shouldWrite) {
-      return;
-    }
-    // Get tracked alert UUIDs to query for
-    // TODO - we can consider refactoring to store the previous execution UUID and query
-    // for active and recovered alerts from the previous execution using that UUID
-    const trackedAlerts = this.legacyAlertsClient.getTrackedAlerts();
-
-    const uuidsToFetch: string[] = [];
-    keys(trackedAlerts).forEach((key) => {
-      const tkey = key as keyof TrackedAlerts<LegacyState, LegacyContext>;
-      keys(trackedAlerts[tkey]).forEach((alertId: string) => {
-        uuidsToFetch.push(trackedAlerts[tkey][alertId].getUuid());
-      });
-    });
-
-    if (!uuidsToFetch.length) {
-      return;
-    }
-
-    const queryByUuid = async (uuids: string[]) => {
+    const getTrackedAlerts = async () => {
       const result = await this.search({
-        size: uuids.length,
         seq_no_primary_term: true,
         query: {
           bool: {
-            filter: [
-              {
-                term: {
-                  [ALERT_RULE_UUID]: this.options.rule.id,
-                },
+            must: {
+              term: {
+                [ALERT_RULE_UUID]: this.options.rule.id,
               },
-              {
-                terms: {
-                  [ALERT_UUID]: uuids,
-                },
+            },
+            must_not: {
+              term: {
+                [ALERT_PREVIOUS_ACTION_GROUP]: ALERT_STATUS_RECOVERED,
```

Comment:
what about alerts that recover a few times in a row (for flapping purposes)? I thought we kept track of them for a few executions, but I could be wrong.

Reply:
AFAIK we convert the recovered alerts to active as long as they are flapping.
And once they recover and becomes active again we create  new alert.
So there shouldn't be any recovered alert that was recovered in the previous execution.
But it is better to get some help from @doakalexi here...

Reply:
And maybe we can improve this PR to add execution UUID to task state and use it to fetch tracked alerts as stated in the comment on line 164 above.

Reply:
yea, i think execution UUID would be good to query off and it seems like something minimal to store in the task state. maybe we can have this query as a fallback in case we can't get the execution UUID for some reason.

Reply:
Yeah that was what I was thinking as well, because there won't be execution uuid initially.
We can remove the fallback once we are sure that all the task states has the uuid.
Maybe with the next major version change.

Reply:
We do need to keep track of the alerts that are recovered for the flapping history. We had an [optimization](https://github.com/elastic/kibana/pull/212454/files#diff-3e0f2d076f01374d4c28b4f1918df508f4bb981e4bf2862fc667925ab5ad5d6dL36) that I had removed in my refactor that we could maybe do here.

Reply:
What would be the previous action group of a flapping recovered alert?
Would it be what ever the active group be or recovered?
Because this query assumes that if a an alert's `previous_action_group` is recovered, it cannot be from the previous execution.

Reply:
We need to still track recovered alerts that aren't from the previous execution. I tested locally using a functional test see what is happening and we need to get docs where `previous_action_group` is recovered in example like this: 

If alert pattern is active alert -> recovered-> recovered  -> active. When the alert is active again it is considered a new alert but we need the flapping history from the last 3 runs. The way you have it now we wouldn't get the flapping history from the second time the alert was recovered bc it's `previous_action_group` is recovered, the flapping history for the new alert would be `[true]` instead of `[true,true,false,true]`.

Reply:
I changed this, now It uses the execution uuid that we save in the task state. And the old query that uses alert ids from the state is there as a fallback. That fallback will also do the migration for us. the fist execution with this branch will use the alert ids and save the execution uuid. Then from the second execution on the execution uuid will be used.